src/main/scala/is/hail/io/bgen/BgenWriter.scala

```diff
@@ -0,0 +1,179 @@
+package is.hail.io.bgen
+
+import is.hail.annotations._
+import is.hail.utils._
+import is.hail.variant.{Genotype, Variant, VariantDataset}
+
+object BgenWriter {
+  val ploidy: Byte = 2
+  val phased: Byte = 0
+
+  def shortToBytesLE(bb: ArrayBuilder[Byte], i: Int) {
+    bb += (i & 0xff).toByte
+    bb += ((i >>> 8) & 0xff).toByte
+  }
+
+  def intToBytesLE(bb: ArrayBuilder[Byte], i: Int) {
+    bb += (i & 0xff).toByte
+    bb += ((i >>> 8) & 0xff).toByte
+    bb += ((i >>> 16) & 0xff).toByte
+    bb += ((i >>> 24) & 0xff).toByte
+  }
+
+  def stringToBytesWShortLength(bb: ArrayBuilder[Byte], s: String) {
+    val bytes = s.getBytes
+    shortToBytesLE(bb, bytes.length)
+    bytes.foreach(bb += _)
+  }
+
+  def stringToBytesWIntLength(bb: ArrayBuilder[Byte], s: String) {
+    val bytes = s.getBytes
+    intToBytesLE(bb, bytes.length)
+    bytes.foreach(bb += _)
+  }
+
+  def bitPack(input: Array[Int], nBitsPerProb: Int): Array[Byte] = {
+    val bb = new ArrayBuilder[Byte]
+    bitPack(bb, input, nBitsPerProb)
+    bb.result()
+  }
+
+  def bitPack(bb: ArrayBuilder[Byte], input: Array[Int], nBitsPerProb: Int) {
+    require(nBitsPerProb >= 1 && nBitsPerProb <= 32, "Number of bits per probability must be between 1 and 32 inclusive.")
+
+    val bitMask = (1L << nBitsPerProb) - 1
+    var byteIndex = 0
+    var data = 0L
+    var dataSize = 0
+
+    input.foreach { i =>
+      data |= ((i.toUIntFromRep.toLong & bitMask) << dataSize)
+      dataSize += nBitsPerProb
+
+      while (dataSize >= 8) {
+        bb += (data & 0xffL).toByte
+        data = data >>> 8
+        dataSize -= 8
+        byteIndex += 1
+      }
+    }
+
+    if (dataSize > 0)
+      bb += (data & 0xffL).toByte
+  }
+
+  def resizeProbInts(input: Array[Int], conversionFactor: Double): Array[Int] = {
+    val resized = input.map(i => i.toUIntFromRep * conversionFactor).zipWithIndex.map { case (d, i) => (i, d, math.floor(d)) }
+    val totalFractional = resized.map(d => d._2 - d._3).sum
+
+    resized.sortBy(d => -(d._2 - d._3)).zipWithIndex.map { case (r, i) =>
+      if (i < totalFractional)
+        (r._1, math.ceil(r._2))
+      else
+        (r._1, math.floor(r._2))
+    }.sortBy(_._1).map(_._2.toUInt.intRep)
```

Comment:
Input size can be greater than 3 -- we support multiallelic variants, but UK Biobank will only be biallelic. I'll make two versions -- one general and one hardcode for input size equal to 3.